File: src/main/java/com/lancer/flink/java/source/SourceApp.java
package com.lancer.flink.java.source;

import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple2;

import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;


import org.apache.flink.util.Collector;
import org.apache.flink.util.NumberSequenceIterator;

import java.util.ArrayList;
import java.util.List;


/**
 * 并行度通过ExecutionConfig类来设定
 *
 * 通过run方法Starts the source ---> 源操作（一个并行度执行一次）
 *
 * 通过open方法，在每个Function使用前做操作  ---> 一个并行度执行一次
 */
public class SourceApp {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        env.execute("SourceApp");
    }


    /*
     * 通过读取ExecutionConfig中的parallelism，通过FileInputFormat读取文件，按parallelism数逻辑切分?
     * */
    public static void test04(StreamExecutionEnvironment env) {
        DataStreamSource<String> source = env.readTextFile("./data/access.log");
        System.out.println(source.getParallelism()); // 16  ?
    }


    // fromParallelCollection
    public static void test02(StreamExecutionEnvironment env) {
        // 调用AbstractRichFunction的open方法，从RuntimeContext中获取task数量（并行度），然后返回一个切分后的迭代器iterator，通过ExecutionConfig中的parallelism来设定调用次数
        /*
        * FromSplittableIteratorFunction中的opne方法：
        * public void open(Configuration parameters) throws Exception {
                int numberOfSubTasks = this.getRuntimeContext().getNumberOfParallelSubtasks(); // 获取task的数量
                int indexofThisSubTask = this.getRuntimeContext().getIndexOfThisSubtask();  // 获取该task的下标
                this.iterator = this.fullIterator.split(numberOfSubTasks)[indexofThisSubTask]; // 返回迭代器数组中，task下标处的迭代器
                this.isRunning = true;
         }
        * */
        DataStreamSource<Long> source = env.fromParallelCollection(new NumberSequenceIterator(1, 10), Long.class);
        System.out.println("fromParallelCollection's parallelism：" + source.getParallelism()); // 16

        SingleOutputStreamOperator<Long> filterStream = source.filter(s -> s >= 5);

        System.out.println("filterStream's parallelism：" + filterStream.getParallelism());
    }
}

File: src/main/java/com/lancer/flink/scala/source/SourceApp.scala
package com.lancer.scala.source

import com.lancer.scala.transformations.Access
import org.apache.flink.api.common.functions.MapFunction
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.functions.source.{RichParallelSourceFunction, RichSourceFunction, SourceFunction}
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer
import org.apache.flink.util.{Collector, NumberSequenceIterator}

import java.util.Properties
import scala.util.Random


object SourceApp {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    test08(env)

    env.execute(this.getClass.getSimpleName)
  }

  def test09(env: StreamExecutionEnvironment): Unit = {
    val source: DataStream[Long] = env.generateSequence(1, 100)

    source.print()
  }

  def test08(env: StreamExecutionEnvironment): Unit = {
    val source: DataStream[String] = env.fromElements[String]("a b c", "a b c")

    println(source.getParallelism) // 1

    source
      .flatMap((line: String, ctx: Collector[String]) => line.split(" ").foreach(ctx.collect))
      .map(new MapFunction[String, (String, Int)] {
        override def map(value: String): (String, Int) = (value, 1)
      })
      .keyBy(0)
      .reduce((t1, t2) => (t1._1, t1._2 + t2._2))
      .print
  }

  def test07(env: StreamExecutionEnvironment): Unit = {
    class AccessSourceV2 extends RichParallelSourceFunction[Access] {
      var running = true

      override def open(parameters: Configuration): Unit = println("~~~~~~open~~~~~~")

      override def run(ctx: SourceFunction.SourceContext[Access]): Unit = {
        val domains = Array[String]("com.com.lancer.com", "a.com", "b.com")
        while (running) {
          for (_ <- 1 to 10)
            ctx.collect(Access(1234567.toLong, domains(Random.nextInt(domains.length)), Random.nextDouble() + 1000))
          Thread.sleep(5000)
        }
      }

      override def cancel(): Unit = running = false
    }
    val source = env.addSource(new AccessSourceV2)
    println(source.getParallelism)
    source.print()
  }

  def test06(env: StreamExecutionEnvironment): Unit = {
    class AccessSourceV1 extends RichSourceFunction[Access] {

      var running = true

      override def open(parameters: Configuration): Unit = println("~~~~~~open~~~~~~")

      override def run(ctx: SourceFunction.SourceContext[Access]): Unit = {
        val domains = Array[String]("com.com.lancer.com", "a.com", "b.com")
        while (running) {
          for (_ <- 1 to 10)
            ctx.collect(Access(1234567.toLong, domains(Random.nextInt(domains.length)), Random.nextDouble() + 1000))
          Thread.sleep(5000)
        }
      }

      override def cancel(): Unit = running = false
    }
    val source = env.addSource(new AccessSourceV1)
    println(source.getParallelism)
    source.print()
  }

  def test05(env: StreamExecutionEnvironment): Unit = {
    val source = env.fromCollection(Array[Int](1, 2, 3))
    println(source.getParallelism) // 1
    source
      .map(i => i * 2)
      .print
  }

  def test04(env: StreamExecutionEnvironment): Unit = {
    val source = env.readTextFile("./data/access.log")
    println(source.getParallelism) // 16
  }

  def test03(env: StreamExecutionEnvironment): Unit = {
    val properties = new Properties()
    properties.setProperty("bootstrap.servers", "bigdata01:9092,bigdata02:9092,bigdata03:9092,bigdata04:9092,bigdata05:9092")
    properties.setProperty("group.id", "test")
    val streamSource = env.addSource(new FlinkKafkaConsumer[String]("flinkTopic", new SimpleStringSchema(), properties))

    println("streamSource's parallelism：" + streamSource.getParallelism)
    streamSource.print()
  }

  def test02(env: StreamExecutionEnvironment): Unit = {
    val source = env.fromParallelCollection(new NumberSequenceIterator(1, 10))
    println("fromParallelCollection's parallelism：" + source.getParallelism) // 16

    val filterStream = source.filter(s => s >= 5)

    println("filterStream's parallelism：" + filterStream.getParallelism)
  }

  def test01(env: StreamExecutionEnvironment): Unit = {
    env.setParallelism(5); // 设置全局并行度

    // 底层都是使用StreamExecutionEnvironment.addSource(new xxxFunction() ---> who implements xxxSourceFunction<T>)
    // SourceFunction并行度为1（底层使用setParallelism(1)，故由优先级，全局并行度被覆盖），ParallelSourceFunction和RichParallelSourceFunction，可以指定并行度
    val source: DataStream[String] = env.socketTextStream("bigdata01", 9999); // .setParallelism(2); 报错，因为只能为1
    println("socketTextStream's parallelism：" + source.getParallelism)

    // 接收socket过来的数据，一行一个单词，把hello过滤掉
    val filterStream = source
      .filter(s => !s.equalsIgnoreCase("hello"))
      .setParallelism(1)

    println("filterStream's parallelism：" + filterStream.getParallelism)
  }
}
